File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.py
from collections import deque

class Solution:
    def lexicographicallySmallestArray(self, nums, limit):
        sted = sorted(nums)

        gn = 0
        sgnum = {}
        ele = {}

        prev = sted[0]

        for x in sted:
            if x - prev <= limit:
                sgnum[x] = gn
                ele.setdefault(gn, deque()).append(x)
            else:
                gn += 1
                sgnum[x] = gn
                ele.setdefault(gn, deque()).append(x)

            prev = x

        for i in range(len(nums)):
            group = sgnum[nums[i]]
            nums[i] = ele[group].popleft()

        return nums

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        vector<int> sted=nums;
        sort(sted.begin(), sted.end());
        int gn=0;
        unordered_map<int, int> sgnum;
        unordered_map<int, queue<int>> ele;
        int prev=sted[0];int i=0;
        for(int i:sted){
            if(i-prev<=limit) {sgnum[i]=gn;ele[gn].push(i);}
            else {gn++;sgnum[i]=gn;ele[gn].push(i);}
            prev=i;
        }
        for(int i=0;i<nums.size();i++){
            int group=sgnum[nums[i]];
            nums[i]=ele[group].front();
            ele[group].pop();
        }
        return nums;
    }
};